contrib/babelfishpg_tsql/src/hooks.c

```diff
@@ -2867,6 +2879,403 @@ pltsql_detect_numeric_overflow(int weight, int dscale, int first_block, int nume
+			return true;
+	return false;
+}
+
+/*
+ * This function will check whether the first 3 inputs are in any of the format
+ * `DD MON YYY`, `DD YYYY MON`, `MON DD YYYY`, `MON YYYY DD`, `YYYY MON DD`, `YYYY DD MON`
+ * where MON is month in text format then returns true if present.
+ */
+static bool containsInTextMonthFormat(int *ftype, char **field)
+{
+	int count_number = 0, count_string = 0;
+
+	#define SWAP_FIELDS(i, j) \
+		char* temp_field;	\
+		int temp_ftype;	\
+		temp_field = field[i];	\
+		temp_ftype = ftype[i];	\
+		field[i] = field[j];	\
+		field[j] = temp_field;	\
+		ftype[i] = ftype[1];	\
+		ftype[j] = temp_ftype;	\
+
+	for (int i = 0; i < 3; i++)
+	{
+		if (ftype[i] == DTK_NUMBER)
+			count_number++;
+		else if (ftype[i] == DTK_STRING)
+		{
+			/* Check whether the string is valid text month */
+			if (!isTextMonthPresent(field[i]))
+				return false;
+			count_string++;
+		}
+		else
+			return false;
+	}
+	if (count_number == 2 && count_string == 1)
+	{
+		/*
+		 * If the first field and third field is an string then swap with the second field
+		 * as when the date is given separatly then all different forms of
+		 * dates is supported. To avoid the conversion failure from `isTextMonthPresent`
+		 * later we are swapping earlier. And convert to Supported date formats which are
+		 * `mm{-/.}dd{-/.}yyyy`, `yyyy{-/.}mm{-/.}dd`
+		 *
+		 * For example convert the input from "JULY 23 2000" to "23 JULY 2000"
+		 * and converting the input from "23 2000 JULY" to "23 2000 JULY"
+		 */
+		if (ftype[0] == DTK_STRING)
+		{
+			SWAP_FIELDS(0, 1);
+		}
+		else if (ftype[2] == DTK_STRING)
+		{
+			SWAP_FIELDS(1, 2);
+		}
+		return true;
+	}
+	return false;
+}
+
+/*
+ * pltsql_time_in
+ *		The function modifies the given input into valid form and
+ *		stores the output to result.
+ * Returns
+ * 		false - If the sql_dialect is not an TSQL
+ * 		true - If the sql_dialect is an TSQL
+ */
+bool pltsql_time_in(const char* str, int32 typmod, TimeADT *result)
+{
+	fsec_t		fsec;
+	struct pg_tm tt,
+			   *tm = &tt;
+	int			tz;
+	int			nf;
+	int			dterr;
+	char		workbuf[MAXDATELEN + 1];
+	char	   *field[MAXDATEFIELDS];
+	int			dtype;
+	int			ftype[MAXDATEFIELDS];
+	StringInfo	res;
+	regex_t		time_regex;
+	char	   *pattern;
+	int len_str = strlen(str);
+	char* modified_str = (char*) malloc(len_str + 1);
+	int j = 0;
+	char* temp_field;
+
+	/* If sql_dialect is not an TSQL then return */
+	if (sql_dialect != SQL_DIALECT_TSQL)
+		return false;
+
+	/* Throw a common error message while casting to time datatype */
+	#define TIME_IN_ERROR()	\
+		ereport(ERROR,	\
+				(errcode(ERRCODE_INVALID_DATETIME_FORMAT),	\
+				 errmsg("Conversion failed when converting date and/or time from character string.")));	\
+
+	/* Remove the spaces present before and after of ':', '/', '.', '-' */
+	for (int i = 0; i < len_str; i++)
+	{
+		if (str[i] == ' ' && (str[i+1] == ' ' || str[i+1] == ':' || str[i+1] == '/' || str[i+1] == '.' || str[i+1] == '-'))
```

Comment:
In addition to the above, we should also check if `i+1 < len_str`

Reply:
Modified the logic to while loop in next revision